Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp

```diff
@@ -1825,14 +1825,33 @@ void MediaPlayerPrivateGStreamer::handleMessage(GstMessage* message)
         asyncStateChangeDone();
         break;
     case GST_MESSAGE_STATE_CHANGED: {
-        if (!messageSourceIsPlaybin || m_isDelayingLoad)
-            break;
-
         GstState newState;
         gst_message_parse_state_changed(message, &currentState, &newState, nullptr);
-
         GST_DEBUG_OBJECT(pipeline(), "Changed state from %s to %s", gst_element_state_get_name(currentState), gst_element_state_get_name(newState));
 
+#if USE(GSTREAMER_HOLEPUNCH)
+        if (currentState == GST_STATE_NULL && newState == GST_STATE_READY) {
+            // If we didn't create a video sink, store a reference to the created one.
+            if (!m_videoSink) {
+                // Detect the videoSink element. Getting the video-sink property of the pipeline requires
+                // locking some elements, which may lead to deadlocks during playback. Instead, identify
+                // the videoSink based on its metadata.
+                GstElement* element = GST_ELEMENT(GST_MESSAGE_SRC(message));
+                if (GST_IS_BASE_SINK(element)) {
+                    const gchar* klass_str = gst_element_get_metadata(element, "klass");
+                    if (strstr(klass_str, "Sink") && strstr(klass_str, "Video")) {
+                        m_videoSink = element;
+
+                        // Ensure that there's a buffer with the transparent rectangle available when playback is going to start.
+                        pushNextHolePunchBuffer();
+                    }
+                 }
+            }
+        }
+#endif
+        if (!messageSourceIsPlaybin || m_isDelayingLoad)
+            break;
+
```

Comment:
I think [the code to dump the pipeline](https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/a0920126bfc883f2a8cf73db95170a3226c5cc9a#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR1973) is still very useful. I would add it here, after the break.